### PR TITLE
Prompt and delayed nu-fission group-to-group matrices

### PR DESCRIPTION
**Pull request by samuelshaner**
This PR extends PR #739 by implementing group-to-group matrix MGXS objects and group-to-group matrix inputs for prompt and delayed nu-fission for the MG mode solver.

### CODE REVIEW

**Comment by nelsonag** [comment]
I just reviewed and am OK with the changes. I will also give a day for someone else to chime in.

File: openmc/mgxs/mdgxs.py [hunk 1]
```diff
@@ -1733,3 +1737,614 @@ def xs_tally(self):
...
+    """
+
+    @property
+    def filters(self):
+        # Create the non-domain specific Filters for the Tallies
+        group_edges = self.energy_groups.group_edges
+        energy = openmc.EnergyFilter(group_edges)
+        energyout = openmc.EnergyoutFilter(group_edges)
+
+        if self.delayed_groups != None:
```
Comment (paulromano) [inline openmc/mgxs/mdgxs.py]: should be `if self.delayed_groups is not None:`

File: openmc/mgxs/mdgxs.py [hunk 2]
```diff
@@ -1733,3 +1737,614 @@ def xs_tally(self):
...
+            for nuclide in nuclides:
+
+                # Build header for nuclide type
+                if xs_type != 'sum':
+                    string += '{0: <16}=\t{1}\n'.format('\tNuclide', nuclide)
+
+                # Build header for cross section type
+                string += '{0: <16}\n'.format(xs_header)
+
+                if self.delayed_groups != None:
```
Comment (paulromano) [inline openmc/mgxs/mdgxs.py]: should be `if self.delayed_groups is not None:`

File: openmc/mgxs/mdgxs.py [hunk 3]
```diff
@@ -1733,3 +1737,614 @@ def xs_tally(self):
...
+                        string += template.format('', delayed_group)
+                        string += '\n'
+
+                        template = '{0: <12}Group {1} -> Group {2}:\t\t'
+
+                        # Loop over incoming/outgoing energy groups ranges
+                        for in_group in range(1, self.num_groups + 1):
+                            for out_group in range(1, self.num_groups + 1):
+                                string += template.format('', in_group, out_group)
+                                average = \
```
Comment (paulromano) [inline openmc/mgxs/mdgxs.py]: Not clear why a line continuation is used when it doesn't save any horizontal space?

File: openmc/mgxs/mdgxs.py [hunk 4]
```diff
@@ -1733,3 +1737,614 @@ def xs_tally(self):
...
+                                          self.get_xs([in_group], [out_group],
+                                                      [subdomain], [nuclide],
+                                                      xs_type=xs_type,
+                                                      value='mean',
+                                                      delayed_groups=[delayed_group])
+                                rel_err = \
```
Comment (paulromano) [inline openmc/mgxs/mdgxs.py]: Same thing here

File: openmc/mgxs/mdgxs.py [hunk 5]
```diff
@@ -1733,3 +1737,614 @@ def xs_tally(self):
...
+                                          self.get_xs([in_group], [out_group],
+                                                      [subdomain], [nuclide],
+                                                      xs_type=xs_type,
+                                                      value='rel_err',
+                                                      delayed_groups=[delayed_group])
+                                average = average.flatten()[0]
+                                rel_err = rel_err.flatten()[0] * 100.
+                                string += '{:1.2e} +/- {:1.2e}%'.format(average,
```
Comment (paulromano) [inline openmc/mgxs/mdgxs.py]: The format specifiers give a minimum width of 1, which is not very useful considering the precision is 2. Consider changing to `{:.2e}`

File: openmc/mgxs/mdgxs.py [hunk 6]
```diff
@@ -1733,3 +1737,614 @@ def xs_tally(self):
...
+                        string += '\n'
+                else:
+
+                    template = '{0: <12}Group {1} -> Group {2}:\t\t'
+
+                    # Loop over incoming/outgoing energy groups ranges
+                    for in_group in range(1, self.num_groups + 1):
+                        for out_group in range(1, self.num_groups + 1):
+                            string += template.format('', in_group, out_group)
+                            average = \
```
Comment (paulromano) [inline openmc/mgxs/mdgxs.py]: no need for line continuation

File: openmc/mgxs/mdgxs.py [hunk 7]
```diff
@@ -1733,3 +1737,614 @@ def xs_tally(self):
...
+                                      self.get_xs([in_group], [out_group],
+                                                  [subdomain], [nuclide],
+                                                  xs_type=xs_type, value='mean')
+                            rel_err = \
```
Comment (paulromano) [inline openmc/mgxs/mdgxs.py]: no need for line continuation

File: openmc/mgxs/mdgxs.py [hunk 8]
```diff
@@ -1733,3 +1737,614 @@ def xs_tally(self):
...
+                                      self.get_xs([in_group], [out_group],
+                                                  [subdomain], [nuclide],
+                                                  xs_type=xs_type, value='rel_err')
+                            average = average.flatten()[0]
+                            rel_err = rel_err.flatten()[0] * 100.
+                            string += '{:1.2e} +/- {:1.2e}%'.format(average,
```
Comment (paulromano) [inline openmc/mgxs/mdgxs.py]: format specifier comment as above

File: openmc/mgxs/mdgxs.py [hunk 9]
```diff
@@ -1733,3 +1737,614 @@ def xs_tally(self):
...
+            if self.domain_type == 'distribcell':
+                string += \
+                    '{0: <16}=\t{1}\n'.format('\tSubdomain', subdomain)
+
+            # Loop over all Nuclides
+            for nuclide in nuclides:
+
+                # Build header for nuclide type
+                if xs_type != 'sum':
+                    string += '{0: <16}=\t{1}\n'.format('\tNuclide', nuclide)
```
Comment (paulromano) [inline openmc/mgxs/mdgxs.py]: A general comment (this applies many places): we are definitely Python 2.7 dependent, so there is no need to explicitly number the fields, i.e. `'{: <16}=\t{}\n'` would do just as well. Prefer not using numbered fields unless explicitly needed.

File: openmc/mgxs_library.py [hunk 10]
```diff
@@ -1086,12 +1092,13 @@ def set_nu_fission_mgxs(self, nu_fission, temperature=294.,
     def set_prompt_nu_fission_mgxs(self, prompt_nu_fission, temperature=294.,
                                    nuclide='total', xs_type='macro',
                                    subdomain=None):
-        """This method allows for an openmc.mgxs.PromptNuFissionXS to be used to
-        set the prompt-nu-fission cross section for this XSdata object.
+        """This method allows for an openmc.mgxs.PromptNuFissionXS or
+        openmc.mgxs.PromptNuFissionMatrixXS to be used to set the
+        prompt-nu-fission cross section for this XSdata object.
```
Comment (paulromano) [inline openmc/mgxs_library.py]: Docstrings should have a one-line summary followed by an optional paragraph that elaborates.

File: openmc/mgxs/mdgxs.py [hunk 11]
```diff
@@ -1733,3 +1737,614 @@ def xs_tally(self):
...
+            A string for the type of value to return. Defaults to 'mean'.
+        delayed_groups : list of int or 'all'
+            Delayed groups of interest. Defaults to 'all'.
+        squeeze : bool
+            A boolean representing whether to eliminate the extra dimensions
+            of the multi-dimensional array to be returned. Defaults to True.
+
+        Returns
+        -------
+        ndarray
```
Comment (paulromano) [inline openmc/mgxs/mdgxs.py]: numpy.ndarray

File: openmc/mgxs/mdgxs.py (hunk 9 above)
Comment (samuelshaner) [inline openmc/mgxs/mdgxs.py]: I took the numbers out in this method, but will mention that the rest of the MGXS module uses numbers when formatting strings. I'm happy to remove those, but feel it would be more appropriate to do so in another PR.

**Comment by samuelshaner** [comment]
@paulromano - your comments have been addressed. The items you brought up with regards to numbers in the formatted strings in the MGXS module and the one line summary for the doc-string headers in `mgxs_library.py` are more systemic issues so I just addressed them in the code added for this PR. I'm happy to make further PRs to address those items if you would like them corrected throughout.
